railties/lib/rails/commands/unused_routes/unused_routes_command.rb

```diff
@@ -0,0 +1,71 @@
+# frozen_string_literal: true
+
+require "rails/command"
+
+module Rails
+  module Command
+    class UnusedRoutesCommand < Base # :nodoc:
+      class RouteInfo
+        def initialize(route)
+          requirements = route.requirements
+          @controller_name = requirements[:controller]
+          @action_name = requirements[:action]
+          @controller_class = (@controller_name.to_s.camelize + "Controller").safe_constantize
+        end
+
+        def invalid?
+          controller_class_missing? || (action_missing? && template_missing?)
+        end
+
+        private
+          def view_path(root)
+            File.join(root.path, @controller_name, @action_name)
+          end
+
+          def controller_class_missing?
+            @controller_name && @controller_class.nil?
+          end
+
+          def template_missing?
+            @controller_class && @controller_class.try(:view_paths).to_a.flat_map { |path| Dir["#{view_path(path)}.*"] }.none?
+          end
+
+          def action_missing?
+            @controller_class && @controller_class.instance_methods.exclude?(@action_name.to_sym)
```

Comment:
I think that this should be checking `@controller_class.action_methods` instead. And we might want to check if the controller responds to `action_missing` as well.

Reply:
action_missing I'm a little hesitant about because we don't know _what action_ action_missing is handling, so I wouldn't call a route valid if the controller has that defined. We can definitely try to denote this in the output though!